src/pyclass_init.rs

```diff
@@ -0,0 +1,122 @@
+use crate::pyclass::{PyClass, PyClassShell};
+use crate::type_object::{PyObjectLayout, PyObjectSizedLayout, PyTypeInfo};
+use crate::{PyResult, Python};
+use std::marker::PhantomData;
+
+pub trait PyObjectInit<T: PyTypeInfo>: Sized {
+    fn init_class(self, shell: &mut T::ConcreteLayout);
+}
+
+pub struct PyNativeTypeInitializer<T: PyTypeInfo>(PhantomData<T>);
+
+impl<T: PyTypeInfo> PyObjectInit<T> for PyNativeTypeInitializer<T> {
+    fn init_class(self, _shell: &mut T::ConcreteLayout) {}
+}
+
+/// An initializer for `PyClassShell<T>`.
+///
+/// You can use this type to initalize complicatedly nested `#[pyclass]`.
+///
+/// ```
+/// # use pyo3::prelude::*;
+/// # use pyo3::py_run;
+/// #[pyclass]
+/// struct BaseClass {
+///     #[pyo3(get)]
+///     basename: &'static str,
+/// }
+/// #[pyclass(extends=BaseClass)]
+/// struct SubClass {
+///     #[pyo3(get)]
+///     subname: &'static str,
+/// }
+/// #[pyclass(extends=SubClass)]
+/// struct SubSubClass {
+///     #[pyo3(get)]
+///     subsubname: &'static str,
+/// }
+///
+/// #[pymethods]
+/// impl SubSubClass {
+///     #[new]
+///     fn new() -> PyClassInitializer<Self> {
+///         let base_init = PyClassInitializer::from(BaseClass{basename: "base"});
+///         base_init.add_subclass(SubClass{subname: "sub"})
+///                  .add_subclass(SubSubClass{subsubname: "subsub"})
+///     }
+/// }
+/// let gil = Python::acquire_gil();
+/// let py = gil.python();
+/// let typeobj = py.get_type::<SubSubClass>();
+/// let inst = typeobj.call((), None).unwrap();
+/// py_run!(py, inst, r#"
+///         assert inst.basename == 'base'
+///         assert inst.subname == 'sub'
+///         assert inst.subsubname == 'subsub'"#);
+/// ```
+pub struct PyClassInitializer<T: PyClass> {
+    init: T,
+    super_init: Box<<T::BaseType as PyTypeInfo>::Initializer>,
+}
+
+impl<T: PyClass> PyClassInitializer<T> {
+    pub fn new(init: T, super_init: <T::BaseType as PyTypeInfo>::Initializer) -> Self {
+        Self {
+            init,
+            super_init: Box::new(super_init),
```

Comment:
Is it necessary to box `super_init`? This seems to be an allocation we may not need.

Reply:
You're right, thank you.